### PR TITLE
439 no support for buildx build build context

### LINKED ISSUE #439: No support for buildx build --build-context ...
## Background
https://github.com/docker/buildx/releases/tag/v0.8.0

Build command now accepts --build-context flag to define additional named build contextes for your builds. For example, these can be used expose additional local directories to your build or pin an image to a specific digest. This feature requires using Dockerfile 1.4+ frontend. https://github.com/docker/buildx/pull/904 [Documentation](https://github.com/docker/buildx/blob/v0.8.0/docs/reference/buildx_build.md#build-context)

## Changes
Add --build-context arg for BuildxCLI

```
def build(
        self,
        ...
        build_contexts: Dict[str, str] = {},
    )...
...
full_cmd.add_args_list("--build-context", format_dict_for_cli(build_contexts))
```

### PR DESCRIPTION
**Pull request by bigcat2014**
Fixes #439

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Sorry it seems there is an issue with the CI. I'll try to fix it quickly. Thanks for your patience! In the meantime, could you add a simple unit test for this new feature?

File: python_on_whales/components/buildx/cli_wrapper.py [hunk 1]
```diff
@@ -208,7 +208,7 @@ def build(
         allow: List[str] = [],
         attest: Optional[Dict[str, str]] = None,
         build_args: Dict[str, str] = {},
-        # TODO: build_context
+        build_contexts: Dict[str, str] = {},
```
Comment (gabrieldemarmiesse) [inline python_on_whales/components/buildx/cli_wrapper.py:211]:
```suggestion
        build_contexts: Dict[str, Union[str, ValidPath]] = {},
```
Any path-like object will also be accepted for local directories

**Review by gabrieldemarmiesse — CHANGES_REQUESTED** [review]

**Comment by bigcat2014** [comment]
Working on unit tests at the moment, although this is my first contribution to open source and I've never had to do unit tests before, so it's taking me a min to work through the structure of them. Should be able to have something to review soon though.

**Comment by bigcat2014** [comment]
So, with the current unit tests, 2 of the ones that I added are expected to fail until it is merged into master because they use absolute paths to files that exist on the branch, which will become master once this is merged (references a `test_context` directory that I created in the git repository and a .tar file)

**Comment by bigcat2014** [comment]
@gabrieldemarmiesse Any movement on this recently? Anything I can update in the unit tests to move this along more quickly?

**Comment by gabrieldemarmiesse** [comment]
yeah I'm sorry, this pull request fell at the bottom of the list. Thanks for reminding me. It seems the tests are not trivial so I think I'll handle them myself. I'll likely have some time tomorrow afternoon to work on it. Thanks for your patience!

**Comment by bigcat2014** [comment]
I've updated some of the unit tests, adding a test for oci layout compliant directories and removing the test for .tar files as they're not explicitly supported by the --build-context flag. The only test that's still a little goofy to implement is the git repository one, but I think what I have is at least a good starting point for you!

**Comment by gabrieldemarmiesse** [comment]
thanks I'll try it out :)

**Comment by gabrieldemarmiesse** [comment]
@bigcat2014 do you need help?

**Comment by bigcat2014** [comment]
I'm not sure, I keep running the unit tests locally, they pass, then they don't pass the automated checks, so I'm not sure exactly what's going on. If the current ones don't pass then I may need you to take a look

**Comment by bigcat2014** [comment]
Ok, the unit tests I have added seem to be passing now. I think it was an issue with my understanding of the "with_docker_driver" and "with_container_driver" fixtures. I updated my tests to use what I think are the correct fixtures now and it seems to be working. I think I've also covered all of the --build-context use cases (local path, git repo, oci layout compliant directory, and docker image)

**Comment by bigcat2014** [comment]
Ok, I'm not entirely sure what's going on. I had all of the unit tests passing except the linting, I ran the linter to fix formatting. It only updated the test_buildx_cli_wrapper.py file and only updated formatting of the unit tests I added, but now a different unit test that I did not add, which was not reformatted at all, is failing.

File: tests/python_on_whales/components/buildx/test_buildx_cli_wrapper.py [hunk 2]
```diff
@@ -30,6 +46,25 @@ def with_container_driver():
...
+
+    current_builder = docker.buildx.inspect()
+    with docker.buildx.create(use=True):
+        docker.buildx.build(tmp_path, output={"type": "oci", "dest": tar_path})
+    docker.buildx.use(current_builder)
+
+    # Extract tar to directory
+    tar = tarfile.open(tar_path)
+    tar.extractall(oci_folder_path)
+    tar.close()
```
Comment (gabrieldemarmiesse) [inline tests/python_on_whales/components/buildx/test_buildx_cli_wrapper.py:65]:
```suggestion
    with tarfile.open(tar_path) as tar:
        tar.extractall(oci_folder_path)
```

**Comment by gabrieldemarmiesse** [comment]
We do have some flaky tests here and there, so it's possible that some tests fail randomly :(

**Comment by gabrieldemarmiesse** [comment]
thanks a lot! It's going to help a lot of users :)

**Review by gabrieldemarmiesse — APPROVED** [review]

**Pull request merged**